File: ruby/searchsettings.rb
################################################################################
#
# searchsettings.rb
#
# class SearchSettings: encapsulates search settings
#
################################################################################

class SearchSettings

  attr_accessor :archivesonly
  attr_accessor :debug
  attr_accessor :dotiming
  attr_accessor :excludehidden
  attr_accessor :firstmatch
  attr_accessor :in_archivefilepatterns
  attr_accessor :in_dirpatterns
  attr_accessor :in_extensions
  attr_accessor :in_filepatterns
  attr_accessor :linesafter
  attr_accessor :linesaftertopatterns
  attr_accessor :linesafteruntilpatterns
  attr_accessor :linesbefore
  attr_accessor :listdirs
  attr_accessor :listfiles
  attr_accessor :listlines
  attr_accessor :maxlinelength
  attr_accessor :multilinesearch
  attr_accessor :out_archivefilepatterns
  attr_accessor :out_dirpatterns
  attr_accessor :out_extensions
  attr_accessor :out_filepatterns
  attr_accessor :printresults
  attr_accessor :printusage
  attr_accessor :printversion
  attr_accessor :recursive
  attr_accessor :searcharchives
  attr_accessor :searchpatterns
  attr_accessor :startpath
  attr_accessor :uniquelines
  attr_accessor :verbose

  def initialize
    @archivesonly = true
    @debug = false
    @dotiming = false
    @excludehidden = true
    @firstmatch = false
    @linesafter = 0
    @linesbefore = 0
    @listdirs = false
    @listfiles = false
    @listlines = false
    @maxlinelength = 150
    @multilinesearch = false
    @printresults = true
    @printusage = false
    @printversion = false
    @recursive = true
    @searcharchives = false
    @startpath = nil
    @uniquelines = false
    @verbose = false

    @in_archivefilepatterns = []
    @in_dirpatterns = []
    @in_extensions = []
    @in_filepatterns = []
    @in_linesafterpatterns = []
    @in_linesbeforepatterns = []
    @linesaftertopatterns = []
    @linesafteruntilpatterns = []
    @out_archivefilepatterns = []
    @out_dirpatterns = get_default_out_dirpatterns()
    @out_extensions = []
    @out_filepatterns = get_default_out_filepatterns()
    @out_linesafterpatterns = []
    @out_linesbeforepatterns = []
    @searchpatterns = []
  end

  def get_default_out_dirpatterns()
    out_dirs = ["\\bCVS$", "\\.git$", "\\.svn$"]
    out_dirpatterns = []
    out_dirs.each do |d|
      out_dirpatterns.push(Regexp.new(d))
    end
    out_dirpatterns
  end

  def get_default_out_filepatterns()
    out_files = ["^\\.DS_Store$"]
    out_filepatterns = []
    out_files.each do |f|
      out_filepatterns.push(Regexp.new(f))
    end
    out_filepatterns
  end

  def add_comma_delimited_exts(exts, ext_set)
    exts.split(",").each do |x|
        ext_set.push(x)
    end
  end

  def add_searchpattern(pattern)
    @searchpatterns.push(Regexp.new(pattern, true))
  end

  def to_s
    s = "SearchSettings(startpath: \"#{@startpath}\""
    s += ", searchpatterns: [\"#{@searchpatterns.join("\", \"")}\"]"
    if @in_extensions
      s += ", in_extensions: [#{@in_extensions.join(", ")}]"
    end
    if @out_extensions
      s += ", out_extensions: [#{@out_extensions.join(", ")}]"
    end
    if @in_dirpatterns
      s += ", in_dirpatterns: [#{@in_dirpatterns.join(", ")}]"
    end
    if @out_dirpatterns
      s += ", out_dirpatterns: [#{@out_dirpatterns.join(", ")}]"
    end
    if @in_filepatterns
      s += ", in_filepatterns: [#{@in_filepatterns.join(", ")}]"
    end
    if @out_filepatterns
      s += ", out_filepatterns: [#{@out_filepatterns.join(", ")}]"
    end
    s += ", archivesonly: #{@archivesonly}"
    s += ", debug: #{@debug}"
    s += ", dotiming: #{@dotiming}"
    s += ", firstmatch: #{@firstmatch}"
    s += ", excludehidden: #{@excludehidden}"
    s += ", linesafter: #{@linesafter}"
    s += ", linesbefore: #{@linesbefore}"
    s += ", listdirs: #{@listdirs}"
    s += ", listfiles: #{@listfiles}"
    s += ", listlines: #{@listlines}"
    s += ", maxlinelength: #{@maxlinelength}"
    s += ", multilinesearch: #{@multilinesearch}"
    s += ", printresults: #{@printresults}"
    s += ", printusage: #{@printusage}"
    s += ", printversion: #{@printversion}"
    s += ", recursive: #{@recursive}"
    s += ", searcharchives: #{@searcharchives}"
    s += ", uniquelines: #{@uniquelines}"
    s += ", verbose: #{@verbose}"
    s += ")"
    s
  end
end

File: node/searchsettings.js
/*
 * nodesearch.js
 *
 * file search utility written in node.js
 */

function SearchSettings() {
    var that = this;
    this.DEFAULT_OUT_DIRPATTERNS = [
        new RegExp("\\bCVS$"),
        new RegExp("\\.git$"),
        new RegExp("\\.svn$")]
    this.DEFAULT_OUT_FILEPATTERNS = [new RegExp("^\\.DS_Store$")]
    this.startPath = "";
    this.inExtensions = [];
    this.outExtensions = [];
    this.inDirPatterns = [];
    this.outDirPatterns = this.DEFAULT_OUT_DIRPATTERNS;
    this.inFilePatterns = [];
    this.outFilePatterns = this.DEFAULT_OUT_FILEPATTERNS;
    this.inArchiveFilePatterns = [];
    this.outArchiveFilePatterns = [];
    this.inLinesAfterPatterns = [];
    this.outLinesAfterPatterns = [];
    this.inLinesBeforePatterns = [];
    this.outLinesBeforePatterns = [];
    this.linesAfterToPatterns = [];
    this.linesAfterUntilPatterns = [];
    this.searchPatterns = [];
    this.archivesOnly = false;
    this.debug = false;
    this.doTiming = false;
    this.firstMatch = false;
    this.linesAfter = 0;
    this.linesBefore = 0;
    this.listDirs = false;
    this.listFiles = false;
    this.listLines = false;
    this.multilineSearch = false;
    this.printResults = false;
    this.printUsage = false;
    this.printVersion = false;
    this.searchArchives = false;
    this.uniqueLines = false;
    this.verbose = false;
    var addExtension = function (ext, arr) {
        arr.push(ext);
    };
    this.addInExtension = function (ext) {
        addExtension(ext, that.inExtensions);
    };
    this.addOutExtension = function (ext) {
        addExtension(ext, that.outExtensions);
    };
    var addPattern = function (pattern, arr) {
        arr.push(new RegExp(pattern));
    };
    this.addInDirPattern = function (pattern) {
        addPattern(pattern, that.inDirPatterns);
    };
    this.addOutDirPattern = function (pattern) {
        addPattern(pattern, that.outDirPatterns);
    };
    this.addInFilePattern = function (pattern) {
        addPattern(pattern, that.inFilePatterns);
    };
    this.addOutFilePattern = function (pattern) {
        addPattern(pattern, that.outFilePatterns);
    };
    this.addSearchPattern = function (pattern) {
        addPattern(pattern, that.searchPatterns);
    };
    this.addInArchiveFilePattern = function (pattern) {
        addPattern(pattern, that.inArchiveFilePatterns);
    };
    this.addOutArchiveFilePattern = function (pattern) {
        addPattern(pattern, that.outArchiveFilePatterns);
    };
    this.addInLinesAfterPattern = function (pattern) {
        addPattern(pattern, that.inLinesAfterPatterns);
    };
    this.addOutLinesAfterPattern = function (pattern) {
        addPattern(pattern, that.outLinesAfterPatterns);
    };
    this.addInLinesBeforePattern = function (pattern) {
        addPattern(pattern, that.inLinesBeforePatterns);
    };
    this.addOutLinesBeforePattern = function (pattern) {
        addPattern(pattern, that.outLinesBeforePatterns);
    };

    this.addLinesAfterToPattern = function (pattern) {
        addPattern(pattern, that.linesAfterToPatterns);
    };
    this.addLinesAfterUntilPattern = function (pattern) {
        addPattern(pattern, that.linesAfterUntilPatterns);
    };

    this.toString = function () {
        var s = 'SearchSettings(startPath="' + that.startPath + '"';
        if (that.inExtensions.length) {
            s = s + ', inExtensions=["' + that.inExtensions.join('","') + '"]';
        }
        if (that.outExtensions.length) {
            s = s + ', outExtensions=["' + that.outExtensions.join('","') + '"]';
        }
        if (that.inDirPatterns.length) {
            s = s + ', inDirPatterns=["' + that.inDirPatterns.join('","') + '"]';
        }
        if (that.outDirPatterns.length) {
            s = s + ', outDirPatterns=["' + that.outDirPatterns.join('","') + '"]';
        }
        if (that.inFilePatterns.length) {
            s = s + ', inFilePatterns=["' + that.inFilePatterns.join('","') + '"]';
        }
        if (that.outFilePatterns.length) {
            s = s + ', outFilePatterns=["' + that.outFilePatterns.join('","') + '"]';
        }
        if (that.searchPatterns.length) {
            s = s + ', searchPatterns=["' + that.searchPatterns.join('","') + '"]';
        }
        s = s + ', doTiming=' + that.doTiming + ', verbose=' + that.verbose + ')';
        return s;
    };
}

exports.SearchSettings = SearchSettings;